packages/core/src/typegenTypes.ts

```diff
@@ -0,0 +1,262 @@
+  TResolvedTypesMeta
+> = TResolvedTypesMeta & {
+  missingImplementations: {
+    actions: never;
+    delays: never;
+    guards: never;
+    services: never;
+  };
+};
+
+// we don't even have to do that much here, technically, because `T & unknown` is equivalent to `T`
+// however, this doesn't display nicely in IDE tooltips, so let's fix this
+type MergeWithInternalEvents<TIndexedEvents, TInternalEvents> = TIndexedEvents &
+  // alternatively we could consider using key remapping in mapped types for this in the future
+  // in theory it would be a single iteration rather than two
+  Pick<TInternalEvents, Exclude<keyof TInternalEvents, keyof TIndexedEvents>>;
+
+export type ResolveTypegenMeta<
+  TTypesMeta extends TypegenConstraint,
+  TEvent extends EventObject,
+  TAction extends BaseActionObject
+> = TTypesMeta extends TypegenEnabled
+  ? TTypesMeta & {
+      indexedActions: IndexByType<TAction>;
+      indexedEvents: MergeWithInternalEvents<
+        IndexByType<TEvent>,
+        Prop<TTypesMeta, 'internalEvents'>
+      >;
+    }
+  : TypegenDisabled;
+
+export type TypegenMachineOptionsActions<
+  TContext,
+  TResolvedTypesMeta,
+  TEventsCausingActions = Prop<TResolvedTypesMeta, 'eventsCausingActions'>,
+  TIndexedEvents = Prop<TResolvedTypesMeta, 'indexedEvents'>,
+  TIndexedActions = Prop<TResolvedTypesMeta, 'indexedActions'>
+> = {
+  [K in keyof TEventsCausingActions]?:
+    | ActionObject<
+        TContext,
+        Cast<Prop<TIndexedEvents, TEventsCausingActions[K]>, EventObject>
+      >
+    | ActionFunction<
+        TContext,
+        Cast<Prop<TIndexedEvents, TEventsCausingActions[K]>, EventObject>,
+        Cast<Prop<TIndexedActions, K>, BaseActionObject>
+      >;
+};
+
+export type TypegenMachineOptionsDelays<
+  TContext,
+  TResolvedTypesMeta,
+  TEventsCausingDelays = Prop<TResolvedTypesMeta, 'eventsCausingDelays'>,
+  TIndexedEvents = Prop<TResolvedTypesMeta, 'indexedEvents'>
+> = {
+  [K in keyof TEventsCausingDelays]?: DelayConfig<
+    TContext,
+    Cast<Prop<TIndexedEvents, TEventsCausingDelays[K]>, EventObject>
+  >;
+};
+
+export type TypegenMachineOptionsGuards<
+  TContext,
+  TResolvedTypesMeta,
+  TEventsCausingGuards = Prop<TResolvedTypesMeta, 'eventsCausingGuards'>,
+  TIndexedEvents = Prop<TResolvedTypesMeta, 'indexedEvents'>
+> = {
+  [K in keyof TEventsCausingGuards]?: ConditionPredicate<
+    TContext,
+    Cast<Prop<TIndexedEvents, TEventsCausingGuards[K]>, EventObject>
+  >;
+};
+
+export type TypegenMachineOptionsServices<
+  TContext,
+  TResolvedTypesMeta,
+  TEventsCausingServices = Prop<TResolvedTypesMeta, 'eventsCausingServices'>,
+  TIndexedEvents = Prop<TResolvedTypesMeta, 'indexedEvents'>,
+  TInvokeSrcNameMap = Prop<TResolvedTypesMeta, 'invokeSrcNameMap'>
+> = {
+  [K in keyof TEventsCausingServices]?: InvokeCreator<
+    TContext,
+    Cast<Prop<TIndexedEvents, TEventsCausingServices[K]>, EventObject>,
+    Prop<Prop<TIndexedEvents, Prop<TInvokeSrcNameMap, K>>, 'data'>
+  >;
```

Comment:
There's some strangeness here that's causing this:

https://github.com/statelyai/registry/pull/97

Reply:
Could be the same stuff as below, ignore until I've had a chance to look.